Zigbee: Add dimmable light endpoint class

FaBjE opened the pull request:
## Description of Change
This PR adds a Zigbee endpoint for a normal dimmable light. (So no colors or white-temperature change)
I based it on the "color dimmable light" example.

Unfortunately this endpoint type is not in the ESP Zigbee library by default.
So I had to make my own "esp_zb_dimmable_light_clusters_create" function.
It appears to be functioning good, but please do a proper check if I figured it out correctly.

## Tests scenarios
I've tested this with Home Assistant.
The Arduino applications runs on a ESP32-C6 with Arduino-esp32 core on "master" branch.

I'm looking forward to your feedback 😃

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=10727) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=10727) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/arduino-esp32?pullRequest=10727) it.</sub>

me-no-dev commented:
Also please target the `release/v3.1.x` branch

SuGlider commented:
> [![CLA assistant check](https://camo.githubusercontent.com/bb14f9741a98be03eb447ca21a4a5e96201df401e5be70f9ef52130f4e1ff903/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=10727) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/arduino-esp32?pullRequest=10727) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/arduino-esp32?pullRequest=10727) it.

@FaBjE - Please sign the CLA.

P-R-O-C-H-Y commented:
Hi @FaBjE, I am happy to see a PR from community. Overall it's looking good, but I am missing an implementation of setting the light manually. Can you please check a ZigbeeColorDimmableLight you have as a template on the `release/v3.1.x` branch? There was more stuff added recently. https://github.com/espressif/arduino-esp32/blob/7dc1c928137134b6dab14fc8a274476468c87944/libraries/Zigbee/src/ep/ZigbeeColorDimmableLight.h#L24

P-R-O-C-H-Y commented on libraries/Zigbee/src/ep/ZigbeeDimmableLight.h:
```diff
@@ -0,0 +1,103 @@
+ *
+ *
+ */
+typedef struct esp_zb_dimmable_light_cfg_s
+{
+  esp_zb_basic_cluster_cfg_t basic_cfg;       /*!<  Basic cluster configuration, @ref esp_zb_basic_cluster_cfg_s */
+  esp_zb_identify_cluster_cfg_t identify_cfg; /*!<  Identify cluster configuration, @ref esp_zb_identify_cluster_cfg_s */
+  esp_zb_groups_cluster_cfg_t groups_cfg;     /*!<  Groups cluster configuration, @ref esp_zb_groups_cluster_cfg_s */
+  esp_zb_scenes_cluster_cfg_t scenes_cfg;     /*!<  Scenes cluster configuration, @ref esp_zb_scenes_cluster_cfg_s */
+  esp_zb_on_off_cluster_cfg_t on_off_cfg;     /*!<  On off cluster configuration, @ref esp_zb_on_off_cluster_cfg_s */
+  esp_zb_level_cluster_cfg_t level_cfg;       /*!<  Level cluster configuration, @ref esp_zb_level_cluster_cfg_s */
+} esp_zb_dimmable_light_cfg_t;
```
Please rename the type and structure as suggested. Don't forget to replace it on every place its used in cpp file.
```suggestion
typedef struct zigbee_dimmable_light_cfg_s
{
  esp_zb_basic_cluster_cfg_t basic_cfg;       /*!<  Basic cluster configuration, @ref esp_zb_basic_cluster_cfg_s */
  esp_zb_identify_cluster_cfg_t identify_cfg; /*!<  Identify cluster configuration, @ref esp_zb_identify_cluster_cfg_s */
  esp_zb_groups_cluster_cfg_t groups_cfg;     /*!<  Groups cluster configuration, @ref esp_zb_groups_cluster_cfg_s */
  esp_zb_scenes_cluster_cfg_t scenes_cfg;     /*!<  Scenes cluster configuration, @ref esp_zb_scenes_cluster_cfg_s */
  esp_zb_on_off_cluster_cfg_t on_off_cfg;     /*!<  On off cluster configuration, @ref esp_zb_on_off_cluster_cfg_s */
  esp_zb_level_cluster_cfg_t level_cfg;       /*!<  Level cluster configuration, @ref esp_zb_level_cluster_cfg_s */
} zigbee_dimmable_light_cfg_t;
```

FaBjE replied:
Fixed

P-R-O-C-H-Y commented on libraries/Zigbee/src/ep/ZigbeeDimmableLight.h:
```diff
@@ -0,0 +1,103 @@
+ *        Added here as not supported by ESP Zigbee library.
+ *
+ */
+#define ESP_ZB_DEFAULT_DIMMABLE_LIGHT_CONFIG()                                        \
```
Also please rename the default config.
```suggestion
#define ZIGEEE_DEFAULT_DIMMABLE_LIGHT_CONFIG()                                        \
```

FaBjE replied:
Fixed (did use ZIGBEE_)

FaBjE commented:
@me-no-dev 
I added an example and modified the target branch.

I modified my commit message according the bot message.
I can "rename" the branch, but we will probably loose this PR. Please advise.

@P-R-O-C-H-Y 
Thanks for the review. I have:
- Added the missing functions (Sorry copied from 3.0.7)
- Updated the example to 3.1.x standards
- Renamed define/config struct (did change ZIGEEE_ to ZIGBEE_ as I assumed a typo)

P-R-O-C-H-Y commented:
@FaBjE Please add the source file to the CMakeLists.txt file here: https://github.com/espressif/arduino-esp32/blob/7dc1c928137134b6dab14fc8a274476468c87944/CMakeLists.txt#L281

FaBjE commented:
> @FaBjE Please add the source file to the CMakeLists.txt file here:
> 
> https://github.com/espressif/arduino-esp32/blob/7dc1c928137134b6dab14fc8a274476468c87944/CMakeLists.txt#L281

Added. Somehow it just compiled fine without here...

lucasssvaz commented:
> > @FaBjE Please add the source file to the CMakeLists.txt file here:
> > https://github.com/espressif/arduino-esp32/blob/7dc1c928137134b6dab14fc8a274476468c87944/CMakeLists.txt#L281
> 
> Added. Somehow it just compiled fine without here...

This is only needed when compiling as an IDF component.

lucasssvaz commented on libraries/Zigbee/examples/Zigbee_Dimmable_Light/Zigbee_Dimmable_Light.ino:
```diff
@@ -0,0 +1,132 @@
+ *
+ * Please check the README.md for instructions and more detailed description.
+ *
+ * Created by Jan Procházka (https://github.com/P-R-O-C-H-Y/)
```
```suggestion
 * Created by [FaBjE](https://github.com/FaBjE) based on examples by Jan Procházka
```

FaBjE replied:
I left it there on purpose, as the example is basically a copy the color light.
Just with some tiny modifications to fit the dimmable light. It did not seem right to take credit for it.

lucasssvaz commented on libraries/Zigbee/src/ep/ZigbeeDimmableLight.h:
```diff
@@ -0,0 +1,111 @@
+ *        Added here as not supported by ESP Zigbee library.
+ *
+ */
+#define ZIGBEE_DEFAULT_DIMMABLE_LIGHT_CONFIG()                                  \
+  {                                                                             \
+    .basic_cfg =                                                                \
+      {                                                                         \
+        .zcl_version = ESP_ZB_ZCL_BASIC_ZCL_VERSION_DEFAULT_VALUE,              \
+        .power_source = ESP_ZB_ZCL_BASIC_POWER_SOURCE_DEFAULT_VALUE,            \
+      },                                                                        \
+    .identify_cfg =                                                             \
+      {                                                                         \
+        .identify_time = ESP_ZB_ZCL_IDENTIFY_IDENTIFY_TIME_DEFAULT_VALUE,       \
+      },                                                                        \
+    .groups_cfg =                                                               \
+      {                                                                         \
+        .groups_name_support_id = ESP_ZB_ZCL_GROUPS_NAME_SUPPORT_DEFAULT_VALUE, \
+      },                                                                        \
+    .scenes_cfg =                                                               \
+      {                                                                         \
+        .scenes_count = ESP_ZB_ZCL_SCENES_SCENE_COUNT_DEFAULT_VALUE,            \
+        .current_scene = ESP_ZB_ZCL_SCENES_CURRENT_SCENE_DEFAULT_VALUE,         \
+        .current_group = ESP_ZB_ZCL_SCENES_CURRENT_GROUP_DEFAULT_VALUE,         \
+        .scene_valid = ESP_ZB_ZCL_SCENES_SCENE_VALID_DEFAULT_VALUE,             \
+        .name_support = ESP_ZB_ZCL_SCENES_NAME_SUPPORT_DEFAULT_VALUE,           \
+      },                                                                        \
+    .on_off_cfg =                                                               \
+      {                                                                         \
+        .on_off = ESP_ZB_ZCL_ON_OFF_ON_OFF_DEFAULT_VALUE,                       \
+      },                                                                        \
+    .level_cfg =                                                                \
+      {                                                                         \
+        .current_level = ESP_ZB_ZCL_LEVEL_CONTROL_CURRENT_LEVEL_DEFAULT_VALUE,  \
+      },                                                                        \
+  }
```
To avoid breaking the formatting

```suggestion
// clang-format off
#define ZIGBEE_DEFAULT_DIMMABLE_LIGHT_CONFIG()                                  \
  {                                                                             \
    .basic_cfg =                                                                \
      {                                                                         \
        .zcl_version = ESP_ZB_ZCL_BASIC_ZCL_VERSION_DEFAULT_VALUE,              \
        .power_source = ESP_ZB_ZCL_BASIC_POWER_SOURCE_DEFAULT_VALUE,            \
      },                                                                        \
    .identify_cfg =                                                             \
      {                                                                         \
        .identify_time = ESP_ZB_ZCL_IDENTIFY_IDENTIFY_TIME_DEFAULT_VALUE,       \
      },                                                                        \
    .groups_cfg =                                                               \
      {                                                                         \
        .groups_name_support_id = ESP_ZB_ZCL_GROUPS_NAME_SUPPORT_DEFAULT_VALUE, \
      },                                                                        \
    .scenes_cfg =                                                               \
      {                                                                         \
        .scenes_count = ESP_ZB_ZCL_SCENES_SCENE_COUNT_DEFAULT_VALUE,            \
        .current_scene = ESP_ZB_ZCL_SCENES_CURRENT_SCENE_DEFAULT_VALUE,         \
        .current_group = ESP_ZB_ZCL_SCENES_CURRENT_GROUP_DEFAULT_VALUE,         \
        .scene_valid = ESP_ZB_ZCL_SCENES_SCENE_VALID_DEFAULT_VALUE,             \
        .name_support = ESP_ZB_ZCL_SCENES_NAME_SUPPORT_DEFAULT_VALUE,           \
      },                                                                        \
    .on_off_cfg =                                                               \
      {                                                                         \
        .on_off = ESP_ZB_ZCL_ON_OFF_ON_OFF_DEFAULT_VALUE,                       \
      },                                                                        \
    .level_cfg =                                                                \
      {                                                                         \
        .current_level = ESP_ZB_ZCL_LEVEL_CONTROL_CURRENT_LEVEL_DEFAULT_VALUE,  \
      },                                                                        \
  }
// clang-format on
```

FaBjE replied:
Added

P-R-O-C-H-Y commented:
Thank you @FaBjE for adding another endpoint class :)

P-R-O-C-H-Y commented:
PR is now ready to be merged. @FaBjE hope you don't mind that I committed small change to avoid possible conflicts for future esp-zigbee-sdk updates :)

FaBjE commented:
> PR is now ready to be merged. @FaBjE hope you don't mind that I committed small change to avoid possible conflicts for future esp-zigbee-sdk releases :)

Not at all. Makes complete sense, we just missed it during the rename of the others.
I kept the functions prefix the same as in the zigbee stack, but now that is outside of that scope it shouldn't be.

I'm glad it is merged, and hope it will benefit others.
I think the (only) one still missing is the white-temperature tuneable light.
